kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -95,7 +94,7 @@ internal class KotlinJsonAdapter<T>(
 
     // Confirm all parameters are present, optional, or nullable.
     for (i in 0 until constructorSize) {
-      if (values[i] === ABSENT_VALUE && !constructor.parameters[i].isOptional) {
+      if (values.containsKey(i) && !constructor.parameters[i].isOptional) {
```

Comment:
Let's save off the parameter and wrap it in a null-check instead, rather than checking that it contains the key and then looking it up two more times